drivers/video/video_ctrls.c

```diff
@@ -0,0 +1,241 @@
+	case VIDEO_CTRL_TYPE_MENU:
+	case VIDEO_CTRL_TYPE_INTEGER_MENU:
+		if (min > max || def < min || def > max || min < 0) {
+			return -ERANGE;
+		}
+		return 0;
+	default:
+		return 0;
+	}
+}
+
+static inline void set_type_flag(uint32_t id, enum video_ctrl_type *type, uint32_t *flags)
+{
+	*flags = 0;
+
+	switch (id) {
+	case VIDEO_CID_HFLIP:
+	case VIDEO_CID_VFLIP:
+		*type = VIDEO_CTRL_TYPE_BOOLEAN;
+		break;
+	case VIDEO_CID_POWER_LINE_FREQUENCY:
+	case VIDEO_CID_TEST_PATTERN:
+		*type = VIDEO_CTRL_TYPE_MENU;
+		break;
+	case VIDEO_CID_PIXEL_RATE:
+		*type = VIDEO_CTRL_TYPE_INTEGER64;
+		*flags |= VIDEO_CTRL_FLAG_READ_ONLY;
+		break;
+	default:
+		*type = VIDEO_CTRL_TYPE_INTEGER;
+		break;
+	}
+}
+
+int video_ctrl_init(struct video_ctrl *ctrl, struct video_ctrl_handler *hdl, uint32_t id,
+		    int32_t min, int32_t max, uint32_t step, int32_t def)
+{
+	int ret;
+	enum video_ctrl_type type;
+	uint32_t flags;
+
+	set_type_flag(id, &type, &flags);
+
+	/* Sanity checks */
+	if (id == 0 || id >= VIDEO_CID_PRIVATE_BASE) {
+		return -EINVAL;
+	}
+
+	ret = check_range(type, min, max, step, def);
+	if (ret) {
+		return ret;
+	}
+
+	ctrl->handler = hdl;
+
+	ctrl->id = id;
+	ctrl->type = type;
+	ctrl->flags = flags;
+	ctrl->min = min;
+	ctrl->max = max;
+	ctrl->step = step;
+	ctrl->def = def;
+	ctrl->val = def;
+
+	sys_slist_append(&hdl->list, &ctrl->node);
+
+	return 0;
+}
+
+struct video_ctrl *video_ctrl_find(struct video_ctrl_handler *hdl, uint32_t id)
+{
+	struct video_ctrl *ctrl;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&hdl->list, ctrl, node) {
+		if (ctrl->id == id) {
+			return ctrl;
+		}
+	}
+
+	return NULL;
+}
+
+static int video_ctrl_find_by_root_dev(const struct device *dev, uint32_t id,
+				       struct video_ctrl **ctrl)
+{
+	struct video_device *vdev = video_find_vdev(dev);
+
+	if (!vdev) {
+		return -EINVAL;
+	}
+
+	struct video_ctrl_handler *hdl = &vdev->ctrl_handler;
+
+	if (!hdl) {
+		LOG_ERR("Control handler not found\n");
+		return -ENOTSUP;
+	}
+
+	*ctrl = video_ctrl_find(hdl, id);
+	if (!*ctrl) {
+		LOG_ERR("Control id 0x%x not found\n", id);
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+int video_get_ctrl(const struct device *dev, struct video_control *control)
+{
+	struct video_ctrl *ctrl;
+
+	int ret = video_ctrl_find_by_root_dev(dev, control->id, &ctrl);
+
+	if (ret) {
+		return ret;
+	}
+
+	if (ctrl->flags & VIDEO_CTRL_FLAG_WRITE_ONLY) {
+		LOG_ERR("Control id 0x%x is write-only\n", control->id);
+		return -EACCES;
+	}
+
+	control->val = ctrl->val;
+
+	return 0;
+}
+
+int video_set_ctrl(const struct device *dev, struct video_control *control)
+{
+	struct video_ctrl *ctrl;
+
+	int ret = video_ctrl_find_by_root_dev(dev, control->id, &ctrl);
+
+	if (ret) {
+		return ret;
+	}
+
+	if (ctrl->flags & VIDEO_CTRL_FLAG_READ_ONLY) {
+		LOG_ERR("Control id 0x%x is read-only\n", control->id);
+		return -EACCES;
+	}
+
+	if (control->val < ctrl->min || control->val > ctrl->max) {
+		LOG_ERR("Control value is invalid\n");
+		return -EINVAL;
+	}
```

Comment:
Is that not checked during registration? Can this change at runtime? If so, that can become a race condition?